// telegram-mtproto/lib/service/main/state.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

require('eventemitter2');

var _networker = require('../networker');

var _networker2 = _interopRequireDefault(_networker);

var _request = require('./request');

var _request2 = _interopRequireDefault(_request);

var _netMessage = require('../networker/net-message');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

class MapEmitter extends Map {
  constructor(emitter, scope) {
    super();
    this.scope = scope;
    Object.defineProperty(this, 'emitter', {
      enumerable: false,
      configurable: true,
      writable: false,
      value: emitter
    });
  }
  set(key, value) {
    this.emitter.emit(this.scope, { key, value });
    return super.set(key, value);
  }
  delete(key) {
    this.emitter.emit(this.scope, key);
    return super.delete(key);
  }
  //$FlowIssue
  [Symbol.species](prop) {
    return new Map(prop);
  }
}

class State {
  constructor() {
    this.threads = new Map();
    this.requests = new Map();
    this.messages = new Map();
  }

}

exports.default = State;
