// telegram-mtproto/test/config-server-data.js
const fullDev = {
  path    : 'apiw_test1',
  protocol: 'http',
  port    : 80,
  dcList  : [
    { id: 1, host: '149.154.175.10' },
    { id: 2, host: '149.154.167.40' },
    { id: 3, host: '149.154.175.117' }
  ]
}

const fullDevOut = {
  dc1: 'http://149.154.175.10:80/apiw_test1',
  dc2: 'http://149.154.167.40:80/apiw_test1',
  dc3: 'http://149.154.175.117:80/apiw_test1',
}

const fullProd = {
  path    : 'apiw1',
  protocol: 'http',
  port    : 80,
  dcList  : [
    { id: 1, host: '149.154.175.50' },
    { id: 2, host: '149.154.167.51' },
    { id: 3, host: '149.154.175.100' },
    { id: 4, host: '149.154.167.91' },
    { id: 5, host: '149.154.171.5' },
  ]
}

const fullProdOut = {
  dc1: 'http://149.154.175.50:80/apiw1',
  dc2: 'http://149.154.167.51:80/apiw1',
  dc3: 'http://149.154.175.100:80/apiw1',
  dc4: 'http://149.154.167.91:80/apiw1',
  dc5: 'http://149.154.171.5:80/apiw1',
}

module.exports = {
  full: {
    dev: {
      in : fullDev,
      out: fullDevOut
    },
    prod: {
      in : fullProd,
      out: fullProdOut
    }
  }
}
